#Il quesito della Susi
![Quesito della Susi](risorse/susi.png)

In [4]:
#combinazione ABCDEF

#combinazione di cifre più elevata possibile
NUMMAX = 999999
#ad ogni iterazione diminuiamo di 1
#poichè i vincoli indicano che le cifre sono una più alte verso le cifre più significative,
#il numero sarà maggiore di 500000, quindi meglio partire dall'alto

#valori per accedere alle cifre intermedie
SPLIT3 = 1000
SPLIT2 = 100

num = NUMMAX
trovato = False
iterazioni = 0

while num > 0 and not trovato:
	#vincolo: ABC = 3 * DEF
	nDEF = num % SPLIT3
	nABC = (num - nDEF) / SPLIT3
	if nABC == 3 * nDEF:
		#vincolo: CD = 3 * FG
		nEF = num % SPLIT2
		val = num / SPLIT2
		nCD = val % SPLIT2
		if nCD == 3 * nEF:
			trovato = True
		else:
			num = num - 1
	else:
		num = num - 1
	iterazioni = iterazioni + 1

if trovato:
	print num
else:
    #non ci aspettiamo di arrivare qua
	print "Errore: non esiste un numero che rispetti i vincoli!"
print iterazioni

933311
66689


Abbiamo contato il numero di combinazioni che sono state prese in considerazione prima di trovare quella cercata. 
Proviamo a vedere cosa sarebbe successo (in termini di numero di configurazioni che valutiamo) se fossimo partiti da 000000 per arrivare al numero cercato.
(Naturalmente conoscendo la soluzione non e' necessario fare la simulazione, basta fare i conti ... ma cosi' abbiamo anche il codice)

In [5]:
#combinazione di cifre più elevata possibile
NUMMAX = 999999

SPLIT3 = 1000
SPLIT2 = 100

num = 1000
trovato = False
iterazioni = 0
while num <= NUMMAX and not trovato:
	#vincolo: ABC = 3 * DEF
	nDEF = num % SPLIT3
	nABC = (num - nDEF) / SPLIT3
	if nABC == 3 * nDEF:
		#vincolo: CD = 3 * FG
		nEF = num % SPLIT2
		val = num / SPLIT2
		nCD = val % SPLIT2
		if nCD == 3 * nEF:
			trovato = True
		else:
			num = num + 1
	else:
		num = num + 1
	iterazioni = iterazioni + 1

if trovato:
	print num
else:
	print "Errore: non esiste un numero che rispetti i vincoli!"
print iterazioni


933311
932312


932312 analisi rispetto a 66689

Entrambi sono metodi che si basano sul provare tutto (tanto la fatica la fa il calcolatore), metodi di "forza bruta", invece di fare una ricerca più ... mirata. Ma anche in questi casi si può cercare di fare scelte più furbe, come quella di partire dal valore più alto.

Proviamo a usare le informazioni sui vincoli per ridurre il numero di analisi fatte.

Per esempio, si può generare ad ogni iterazione il numero costituito dalle 3 cifre più a destra (DEF). Calcolare le prime tre in base al vincolo (3 volte il numero DEF, ABC = 3 * DEF), quindi verificare se il secondo vincolo è rispettato (3 volte il numero EF, CD = 3 * EF).

In [10]:
#combinazione di partenza sulle 2 cifre meno significative
#ABCDEF
DEFMAX = 999

SPLIT2 = 100
BASE = 10

nDEF = 1
trovato = False
iterazioni = 0
while nDEF < DEFMAX and not trovato:
    #vincolo ABC = 3 * DEF
    nABC = 3 * nDEF
    #vincolo CD = 3 * EF
    nEF = nDEF % SPLIT2
    nCD = (nABC % BASE) * BASE + nDEF / SPLIT2 
    #verifico il vincolo    
    if nCD == 3 * nEF:
        trovato = True
    else:
        nDEF = nDEF + 1
    iterazioni = iterazioni + 1
        
if trovato:
	print num
else:
	print "Errore: non esiste un numero che rispetti i vincoli!"
print iterazioni

933311
311


Si riducono molto le iterazioni, che nel caso pessimo saranno 999 (DEFMAX).
Utilizzando lo stesso approccio, allora, potremmo lavorare prima sull'altro vincolo che coinvolge solo due cifre, quelle più a destra, visto che essendo il numero di due cifre, al più faremo 99 iterazioni.
Partiamo quindi, con il numero costituito dalle 2 cifre più a destra (EF) e calcoliamo la terza e quarta (CD) con il vincolo (3 volte EF, CD = 3 * EF). A questo punto calcoliamo le tre cifre più significative (ABC) con l'altro vincolo (3 volte DEF, ABC = 3 * DEF) e verifichiamo che la cifra C sia la stessa in entrambi i casi.

In [11]:
#combinazione di partenza sulle 2 cifre meno significative
#ABCDEF
EFMAX = 99

SPLIT2 = 100
BASE = 10

nEF = 1
trovato = False
iterazioni = 0
while nEF < EFMAX and not trovato:
    #vincolo CD = 3 * EF
    nCD = 3 * nEF
    #DEF = cifra meno significativa di CD * 100 + EF
    nDEF = (nCD % BASE) * SPLIT2 + nEF
    #vincolo ABC = 3 * DEF
    nABC = 3 * nDEF

    #C deve essere uguale alla cifra più significativa di CD
    if nCD / BASE == nABC % BASE:
        trovato = True
    else:
        nEF = nEF + 1
    iterazioni = iterazioni + 1
        
if trovato:
	print num
else:
	print "Errore: non esiste un numero che rispetti i vincoli!"
print iterazioni
   

933311
11
